src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs

```diff
@@ -241,34 +241,43 @@ private void AddSourceFiles()
             foreach (var sourcePath in PackTargetArgs.SourceFiles.Keys)
             {
                 var projectDirectory = PackTargetArgs.SourceFiles[sourcePath];
-                if (projectDirectory.EndsWith("\\"))
-                {
-                    projectDirectory = projectDirectory.Substring(0, projectDirectory.LastIndexOf("\\"));
-                }
-                var projectName = Path.GetFileName(projectDirectory);
-                string targetPath = Path.Combine(SourcesFolder, projectName);
-                if (sourcePath.Contains(projectDirectory))
-                {
-                    var relativePath = Path.GetDirectoryName(sourcePath).Replace(projectDirectory, string.Empty);
-                    if (relativePath.StartsWith("\\"))
-                    {
-                        relativePath = relativePath.Substring(1, relativePath.Length - 1);
-                    }
-                    if (relativePath.EndsWith("\\"))
-                    {
-                        relativePath = relativePath.Substring(0, relativePath.LastIndexOf("\\"));
-                    }
-                    targetPath = Path.Combine(targetPath, relativePath);
-                }
+                var finalTargetPath = GetTargetPathForSourceFile(sourcePath, projectDirectory);
+
                 var packageFile = new ManifestFile()
                 {
                     Source = sourcePath,
-                    Target = Path.Combine(targetPath, Path.GetFileName(sourcePath))
+                    Target = finalTargetPath
                 };
                 AddFileToBuilder(packageFile);
             }
         }
 
+        public static string GetTargetPathForSourceFile(string sourcePath, string projectDirectory)
```

Comment:
this method is public only to make it unit testable, but I will still add null checks.